mono/metadata/class.c

```diff
@@ -5252,16 +5249,16 @@ can_access_type (MonoClass *access_klass, MonoClass *member_klass)
 		return TRUE;
 
 	case TYPE_ATTRIBUTE_NESTED_PUBLIC:
-		return TRUE;
+		return member_klass_nested_in && can_access_type (access_klass, member_klass_nested_in);
```

Comment:
I don't think this is right.

If you just bring down the original condition, we have:
```
   if (member_klass_nested_in && !can_access (acess_klass, member_klass_nested_in))
     return FALSE;
   else
     return TRUE;
```
which is equivalent to
```c
    return !member_klass_nested_in || can_access_type (access_klass, member_klass_nested_in);
```

Reply:
On second thought, the old code makes no sense.  Your way is better.

Reply:
It's ok to fail visibility checks for broken code. A class can have nested public visibility and not be nested. I think there's even a check for that early on.